engine/src/main/java/org/apache/apex/stram/GroupingManager.java

```diff
@@ -0,0 +1,232 @@
+   *         <b>Note:</b> groupId 0 indicates and independent event, with no
+   *         group
+   */
+  public EventGroupId getEventGroupIdForContainer(String containerId)
+  {
+    EventGroupId groupId = null;
+    if (groupingRequests.get(containerId) != null) {
+      groupId = groupingRequests.get(containerId).getEventGroupId();
+    }
+    return groupId;
+  }
+
+  /**
+   * Returns deploy/undeploy group Id for container This could be a new
+   * container allocated during redeploy process
+   * @param containerId
+   * @return groupId <br/>
+   *         <b>Note:</b> groupId 0 indicates and indipendent event, with no
+   *         group
+   */
+  public EventGroupId getEventGroupIdForAffectedContainer(String containerId)
+  {
+    EventGroupId groupId = getEventGroupIdForContainer(containerId);
+    if (groupId != null) {
+      return groupId;
+    }
+    for (GroupingRequest request : getGroupingRequests().values()) {
+      if (request.getAffectedContainers().contains(containerId)) {
+        groupId = request.getEventGroupId();
+      }
+    }
+    return groupId;
+  }
+
+  /**
+   * Returns grouping groupId for operator which is to undergo deploy. Operators
+   * undergoing deploy for first time will have groupId as 0
+   * @param operatorId
+   * @return groupId <br/>
+   *         <b>Note:</b> groupId 0 indicates and indipendent event, with no
+   *         group
+   */
+  public EventGroupId getEventGroupIdForOperatorToDeploy(int operatorId)
+  {
+    for (GroupingRequest request : getGroupingRequests().values()) {
+      if (request.getOperatorsToDeploy().contains(operatorId)) {
+        return request.getEventGroupId();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Adds operator to deploy. The operator is added to request associated with containerId
+   * @param containerIs
+   * @param operator
+   */
+  public void addOperatorToDeploy(String containerId, PTOperator oper)
+  {
+    GroupingRequest request = getGroupingRequest(containerId);
+    if (request != null) {
+      request.addOperatorToDeploy(oper.getId());
+    }
+  }
+
+  /**
+   * Removes operator from grouping request
+   */
+  public boolean removeOperatorFromGroupingRequest(int operatorId)
+  {
+    for (GroupingRequest request : getGroupingRequests().values()) {
+      if (request.getOperatorsToDeploy().contains((operatorId))) {
+        return request.removeOperatorToDeploy(operatorId);
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Remove groupingRequest from StrAM if it has no more pending operators to deploy
+   * @param containerId
+   */
+  public void removeProcessedGroupingRequests()
+  {
+    for (Entry<String, GroupingRequest> request : groupingRequests.entrySet()) {
+      if (request.getValue().getOperatorsToDeploy().size() == 0
+          && request.getValue().getOperatorsToUndeploy().size() == 0) {
+        LOG.info("Removing for :" + request.getKey());
+        groupingRequests.remove(request.getKey());
```

Comment:
@DT-Priyanka @bhupeshchawda subject to ConcurrentModificationException as in https://builds.apache.org/job/Apex_Core_PR/org.apache.apex$apex-engine/921/consoleFull